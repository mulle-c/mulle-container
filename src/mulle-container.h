//
//  mulle-container.h
//  mulle-container
//
//  Created by Nat! on 02/11/15.
//  Copyright (c) 2015 Nat! - Mulle kybernetiK.
//  Copyright (c) 2015 Codeon GmbH.
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are met:
//
//  Redistributions of source code must retain the above copyright notice, this
//  list of conditions and the following disclaimer.
//
//  Redistributions in binary form must reproduce the above copyright notice,
//  this list of conditions and the following disclaimer in the documentation
//  and/or other materials provided with the distribution.
//
//  Neither the name of Mulle kybernetiK nor the names of its contributors
//  may be used to endorse or promote products derived from this software
//  without specific prior written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
//  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
//  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
//  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
//  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
//  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
//  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
//  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
//  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
//  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
//  POSSIBILITY OF SUCH DAMAGE.
//
#ifndef mulle_container__h__
#define mulle_container__h__

#define MULLE_CONTAINER_VERSION  ((2 << 20) | (0 << 8) | 0)

#include "include.h"

#include "mulle-container-operation.h"
#include "mulle-container-math.h"
#include "mulle-container-callback.h"
#include "mulle-container-callback-global.h"

#include "mulle-prime.h"
#include "mulle-hash.h"

#include "mulle--array.h"
#include "mulle-array.h"

#include "mulle--set.h"
#include "mulle-set.h"

#include "mulle--map.h"
#include "mulle-map.h"

#include "mulle-range.h"
#include "mulle--rangeset.h"

#include "mulle--pointerqueue.h"
#include "mulle-pointerarray.h"
#include "mulle--pointermap.h"
#include "mulle-pointerset.h"
#include "mulle-pointerpairarray.h"

#include "mulle-container-callback.h"
#include "mulle-container-operation.h"

#if MULLE_C11_VERSION < ((3 << 20) | (2 << 8) | 0)
# error "mulle-c11 is too old"
#endif

#if MULLE_ALLOCATOR_VERSION < ((2 << 20) | (0 << 8) | 0)
# error "mulle-allocator is too old"
#endif


#endif /* mulle_container_h */
