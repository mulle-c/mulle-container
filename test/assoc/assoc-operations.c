//
//  assoc-operations.c
//  test-assoc
//
//  Created by Codeon GmbH.
//  Copyright (c) 2024 Mulle kybernetiK.
//  All rights reserved.
//
#ifdef NDEBUG
# undef NDEBUG
#endif
#include <assert.h>

#include <stdio.h>
#include <mulle-container/mulle-container.h>
#include <mulle-testallocator/mulle-testallocator.h>


static void  test_is_equal( void)
{
   struct mulle_assoc                        *assoc1;
   struct mulle_assoc                        *assoc2;
   struct mulle_container_keyvaluecallback   callback;

   callback.keycallback   = mulle_container_keycallback_int;
   callback.valuecallback = mulle_container_valuecallback_int;

   assoc1 = mulle_assoc_create( 0, &callback, NULL, NULL);
   assoc2 = mulle_assoc_create( 0, &callback, NULL, NULL);

   printf( "Empty assocs equal: %d\n", mulle_assoc_is_equal( assoc1, assoc2));

   mulle_assoc_set( assoc1, (void *) 1, (void *) 10);
   mulle_assoc_set( assoc2, (void *) 1, (void *) 10);
   printf( "Same content equal: %d\n", mulle_assoc_is_equal( assoc1, assoc2));

   mulle_assoc_set( assoc2, (void *) 2, (void *) 20);
   printf( "Different content equal: %d\n", mulle_assoc_is_equal( assoc1, assoc2));

   mulle_assoc_destroy( assoc1);
   mulle_assoc_destroy( assoc2);
}


static void  test_index_access( void)
{
   struct mulle_assoc                        *assoc;
   struct mulle_container_keyvaluecallback   callback;
   struct mulle_pointerpair                  pair;

   callback.keycallback   = mulle_container_keycallback_int;
   callback.valuecallback = mulle_container_valuecallback_int;

   assoc = mulle_assoc_create( 0, &callback, NULL, NULL);

   mulle_assoc_set( assoc, (void *) 1, (void *) 10);
   mulle_assoc_set( assoc, (void *) 2, (void *) 20);
   mulle_assoc_set( assoc, (void *) 3, (void *) 30);

   pair = mulle_assoc_get_at_index( assoc, 0);
   printf( "Index 0: key=%ld value=%ld\n", (long) pair.key, (long) pair.value);

   // Replace key=1 at index 0 with key=2,value=20
   // After sorting, key=2 should be at index 1 since keys are [1,2,3]
   mulle_assoc_set_at_index( assoc, 0, (void *) 2, (void *) 20);
   pair = mulle_assoc_get_at_index( assoc, 1);  // Check index 1 where key=2 should be
   printf( "After set index 0: key=%ld value=%ld\n", (long) pair.key, (long) pair.value);

   mulle_assoc_destroy( assoc);
}


static void  run_test( void (*f)( void))
{
   mulle_testallocator_reset();
   (f)();
   mulle_testallocator_reset();
}


int main(int argc, const char * argv[])
{
   run_test( test_is_equal);
   run_test( test_index_access);
   return( 0);
}