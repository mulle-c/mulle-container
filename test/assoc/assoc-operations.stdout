Empty assocs equal: 1
Same content equal: 1
Different content equal: 0
Index 0: key=1 value=10
After set index 0: key=2 value=20
